src/rust/engine/process_execution/src/nailgun/mod.rs

```diff
@@ -86,7 +86,29 @@ fn construct_nailgun_client_request(
     is_nailgunnable,
   } = original_req;
   client_args.insert(0, client_main_class);
-  ExecuteProcessRequest {
+  // The arg file is only materialized to the client workdir, but because we use nailMain, java is running in a
```

Comment:
I completely agree! I think that is a good solution, and I suggested it in https://github.com/pantsbuild/pants/pull/8750#discussion_r353915856. There was some pushback on the idea from @stuhood saying we should instead use a single client directory per server, and not use nailMain. In my view one of the benefits of the hermetic nailgun feature, is the hermetic part being shareable with local execution. So if we can keep using the existing hermetic execution mechanisms instead of having an exception for nailgun hermetic we should. I will move forward with implementing this in preprocessArgs, which means moving it to the other PR, and removing the code here.

Reply:
My suggestion was to not _rely_ on the existence of `nailMain`. Without having a stable client directory, we would only support tools that had been adapted to support nailgun: which is very, very few tools.

Reply:
server cwd == clients cwd is a new mode of hermetic execution where we snapshot and remove the contents of that particular directory after every command. The issue at hand with Zinc can be fixed by looking for arg files in the nailMain cwd, in our particular zinc wrapper.

For tools that don't implement nailMain the client dir must equal the server dir. Should we design that new mode of hermetic execution when we actually want to implement a tool that doesn't have nailMain, or should we do it now in this PR?

Reply:
I think it's reasonable for us to impose the constraint that nailgun'd servers must support being run from arbitrary directories... Yes it's an additional constraint, but it doesn't seem like a terrible one?

Reply:
@illicitonion : It's fine as a stopgap, but it is not fine in general... all tools expect relative paths to be resolvable to the directory they are running in, which is why nailMain exists. Needing to modify all/most tools you nailgun to add a nailMain that rewrites their arguments would be frustrating (and we nailgun a lot of tools).